source/opt/copy_prop_arrays.h

```diff
@@ -0,0 +1,191 @@
+// 1) The source must never be stored to.
+// 2) The target must be stored to exactly once.
+// 3) The store to the target must be a store to the entire array, and be a
+// copy of the entire source.
+// 4) All loads of the target must be dominated by the store.
+//
+// The hard part is keeping all of the types correct.  We do not want to
+// have to do too large a search to update everything, which may not be
+// possible, do we give up if we see any instruction that might be hard to
+// update.
+
+class CopyPropagateArrays : public MemPass {
+ public:
+  const char* name() const override { return "copy-propagate-arrays"; }
+  Status Process(ir::IRContext*) override;
+
+ private:
+  // The class used to identify a particular memory object.  This memory object
+  // will be owned by a particular variable, meaning that the memory is part of
+  // that variable.  It could be the entire variable or a member of the
+  // variable.
+  class MemoryObject {
+   public:
+    // Construction a memory object that is owned by |var_inst|.  The iterator
+    // |begin| and |end| traverse a container of integers that identify which
+    // member of |var_inst| this memory object will represent.  These integers
+    // are interpreted the same way they would be in an |OpCompositeExtract|
+    // instruction.
+    template <class iterator>
+    MemoryObject(ir::Instruction* var_inst, iterator begin, iterator end);
+
+    // Change |this| to now point to the member identify by |access_chain|
+    // (starting from the current member).  The elements in |access_chain| are
+    // interpreted the same as the indicies in the |OpCompositeExtract|
+    // instruction.
+    void GetMember(const std::vector<uint32_t>& access_chain);
+
+    // Change |this| to now represent the first enclosing object to which it
+    // belongs.  (Remove the last element off the access_chain). It is invalid
+    // to call this function if |this| does not represent a member its owner.
+    void GetParent() {
+      assert(IsMember());
+      access_chain_.pop_back();
+    }
+
+    // Returns true if |this| represents a member of its owner, and not the
+    // entire variable.
+    bool IsMember() const { return !access_chain_.empty(); }
+
+    // Returns the number of members in the object represented by |this|.  If
+    // |this| does not represent a composite type, the return value will be 0.
+    uint32_t GetNumberOfMembers();
+
+    // Returns the owning variable that the memory object is contained in.
+    ir::Instruction* GetVariable() const { return variable_inst_; }
+
+    // Returns a vector of integers that can be used to access the specific
+    // member that |this| represents starting from the owning variable.  These
+    // values are to be interpreted the same way the indicies are in an
+    // |OpCompositeExtract| instruction.
+    const std::vector<uint32_t>& AccessChain() const { return access_chain_; }
+
+    // Returns the type id for a pointer to the memory object.
```

Comment:
Can you describe what this function is doing?  At first, it seemed like it was similar to GetPointeeTypeId(), but it isn't and it's not immediately clear what it's trying to do.